### PR TITLE
test(app): Protocol Upload Revamp e2e test case

### PR DESCRIPTION
**Pull request by reachojhaneha**
**Overview**
Add the ability to drive the installed run app for end to end testing of the feature Protocol Upload revamp(PUR).

- Use python, selenium(4), and pytest to execute tests against the installed run app.
- Roughly use the[ Page Object Model](https://www.toolsqa.com/selenium-webdriver/page-object-model/)
- See [Notes in README](https://github.com/Opentrons/opentrons/compare/app-testing-installed-python-selenium?expand=1#diff-c8553976128c5a29097f45c16088ceddb354664d655950da556cd6b8b380a5acR7)

**TestCase Scenarios:** 

- Happy path scenario for PUR - connecting to robot->landing on protocol page->uploading a protocol -> verifying "setup for run page"
- Multiples of a module - verifying that Moam link is available when a protocol with MoaM is uploaded
- Gen 1 pipette - Links related to Gen 1 pipette and no modules are uploaded.
- No modules uploaded testcases
All new code with no dependencies outside of the new app-testing directory.

**Risk assessment**
There is no risk to any other module in the repository. The app-testing folder is self contained.

### CODE REVIEW

File: app-testing/conftest.py [hunk 1]
```diff
@@ -48,3 +48,15 @@ def test_protocols() -> Dict[str, Path]:
...
+@pytest.fixture(scope="session")
+def test_json_protocols() -> Dict[str, Path]:
+    """Provide a fixture with a dictionary of test protocol files."""
+    # build this manually for now
+    return {
+        "protocoluploadjson": Path(
+            Path(__file__).resolve().parent,
+            "files/protocol/python1/upload_protocol.json",
+        ),
+    }
```
Comment (y3rsh) [inline app-testing/conftest.py]: There is no need to create a whole new fixture.  Please add it to the above fixture like
```python
"python1": Path(
            Path(__file__).resolve().parent,
            "files/protocol/python1/test_drive.py",
        ),
"protocoluploadjson": Path(
            Path(__file__).resolve().parent,
            "files/protocol/json1/upload_protocol.json",
        ),
```

File: app-testing/gen/electron/buildflags/buildflags.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+#define BUILDFLAG_INTERNAL_ENABLE_PDF_VIEWER() (1)
+#define BUILDFLAG_INTERNAL_ENABLE_TTS() (1)
+#define BUILDFLAG_INTERNAL_ENABLE_COLOR_CHOOSER() (1)
+#define BUILDFLAG_INTERNAL_ENABLE_ELECTRON_EXTENSIONS() (1)
+#define BUILDFLAG_INTERNAL_ENABLE_BUILTIN_SPELLCHECKER() (1)
+#define BUILDFLAG_INTERNAL_ENABLE_PICTURE_IN_PICTURE() (1)
+#define BUILDFLAG_INTERNAL_ENABLE_WIN_DARK_MODE_WINDOW_UI() (0)
+#define BUILDFLAG_INTERNAL_OVERRIDE_LOCATION_PROVIDER() (0)
+
+#endif  // ELECTRON_BUILDFLAGS_BUILDFLAGS_H_
```
Comment (y3rsh) [inline app-testing/gen/electron/buildflags/buildflags.h]: Is this added by mistake?

File: app-testing/src/menus/protocol_file.py [hunk 3]
```diff
@@ -21,6 +21,7 @@ class ProtocolFile:
         By.XPATH,
         "(//input[contains(normalize-space(@class), 'upload-panel__file_input')])[2]",
     )
+    file_drag_locator: tuple = (By.ID, "file_input")
```
Comment (y3rsh) [inline app-testing/src/menus/protocol_file.py]: All of the `tuple` type annotations will need to be changed.  Mypi corrected my original version of these from `tuple` to `Tuple[str, str]` 
Make sure to run `make check` it should be complaining about these.

File: app-testing/src/pages/module_setup.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
+"""Model for the screen of module setup"""
```
Comment (y3rsh) [inline app-testing/src/pages/module_setup.py]: Need a `.` at the end, `make check` will catch 😊

File: app-testing/tests/protocol_upload_test.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        ot_application.write_config()
+        robots_list = RobotsList(driver)
+        if not robots_list.is_robot_toggle_active(RobotsList.DEV):
+            robots_list.get_robot_toggle(RobotsList.DEV).click()
+        left_menu = LeftMenu(driver)
+        left_menu.click_more_button()
+        protocol_upload = ProtocolUpload(driver)
+        protocol_upload.click_app_left_panel()
+        protocol_upload.click_enable_developer_toggle()
+        protocol_upload.click_enable_pur_feature()
```
Comment (y3rsh) [inline app-testing/tests/protocol_upload_test.py]: Do you need to click this toggle as a part of the test?  If not you may skip this with a configuration change.  All you need are these 2 lines above line [43 ](https://github.com/Opentrons/opentrons/pull/8507/files#diff-d7380cd6709cc0c827d450c487226a16f61cfe9f6d20237ee49e73c1b6e1e77aR43)
```python
# use new protocol tab
ot_application.config["devInternal"] = {"preProtocolFlowWithoutRPC": True}
```

File: app-testing/tests/protocol_upload_test.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+from src.pages.module_setup import ModuleSetup
+from src.pages.labware_setup import LabwareSetup
+from src.menus.robots_list import RobotsList
+from src.menus.protocol_file import ProtocolFile
+from src.driver.drag_drop import drag_and_drop_file
+
+logger = logging.getLogger(__name__)
+
+
+def test_protocol_upload(
```
Comment (y3rsh) [inline app-testing/tests/protocol_upload_test.py]: Need a docstring describing the test.

File: app-testing/tests/protocol_upload_test.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    chrome_options: Options,
+    test_json_protocols: Dict[str, Path],
+    request: FixtureRequest,
+) -> None:
```
Comment (y3rsh) [inline app-testing/tests/protocol_upload_test.py]: This test requires a robot running or the dev emulator correct?   Including the below validates it is running so the test will fail fast if it is not.
````python
    robot = OtRobot()
    # expecting docker emulated robot
    assert robot.is_alive(), "is a robot available?"
````

File: app-testing/src/pages/robot_calibration.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
+"""Model for the screen of robot calibration."""
+from enum import Enum
```
Comment (y3rsh) [inline app-testing/src/pages/robot_calibration.py]: Unused import, again `make check` will call these out 😊

**Review by y3rsh — CHANGES_REQUESTED** [review]
# 🎉
### Awesome work @reachojhaneha only a little cleanup remaining.

File: app-testing/tests/protocol_upload_test.py (hunk 5 above)
Comment (reachojhaneha) [reply app-testing/tests/protocol_upload_test.py]: Yes, i have to toggle the PUR feature , once it is out in production, i will remove these 3 lines.

File: app-testing/src/pages/moam_pur.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+from selenium.webdriver.common.by import By
+from selenium.webdriver.remote.webelement import WebElement
+from selenium.webdriver.support.wait import WebDriverWait
+from selenium.webdriver.support import expected_conditions as EC
+
+from src.driver.highlight import highlight
+
+logger = logging.getLogger(__name__)
+
+"""Class for Moam Setup flow."""
```
Comment (y3rsh) [inline app-testing/src/pages/moam_pur.py]: Please move these lines in your page objects directly under the class declaration so it becomes a docstring.

File: app-testing/src/pages/module_setup.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        scroll: WebElement = WebDriverWait(self.driver, 2).until(
+            EC.element_to_be_clickable(ModuleSetup.module_setup_text_locator)
+        )
+        actions = ActionChains(self.driver)
+        actions.move_to_element(scroll).perform()
+        return scroll
+
+    @highlight
+    def get_magetic_module(self) -> WebElement:
+        scroll: WebElement = WebDriverWait(self.driver, 2).until(
```
Comment (y3rsh) [inline app-testing/src/pages/module_setup.py]: `scroll` is a confusing name for the element to scroll to in this context.  I thought for a second we were grabbing the scroll bar or something.  I am cool with a generic short name here but I think `mag_module` `target` or `element` would be more optimal.

File: app-testing/src/pages/module_setup.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+            EC.element_to_be_clickable(ModuleSetup.magetic_module)
+        )
+        actions = ActionChains(self.driver)
+        actions.move_to_element(scroll).perform()
+        return scroll
+
+    @highlight
+    def get_proceed_to_labware_setup(self) -> WebElement:
```
Comment (y3rsh) [inline app-testing/src/pages/module_setup.py:89]: A lot of the functions are missing a docstring.  Hopefully `make check` is flagging this.

**Review by y3rsh — CHANGES_REQUESTED** [review]
Looks great but for a few small changes to meet linting.  Try to use the docstrings to describe the purpose of the element from the user perspective.  The code documents clearly the how of the locating.  docstrings for functions that  already have descriptive name are awkward... I suggest be brief `"""Click module."""`.  Most of these wont be helpful but we want consistency.  If there is any unique behavior or state dependency for an element, definitely mention that.

**Review by b-cooper — APPROVED** [review]
Looks good to me 🧪

File: app-testing/files/protocol/python1/moam_pur.json [hunk 12]
```diff
@@ -4,7 +4,7 @@
     "author": "AA BB",
```
Comment (shlokamin) [inline app-testing/files/protocol/python1/moam_pur.json:4]: how come this json file is in a folder called `python1`?

File: app-testing/tests/protocol_upload_test.py [hunk 13]
```diff
@@ -192,6 +188,97 @@ def test_moam_pur(
...
+        ot_application = OtApplication(
+            Path(f"{driver.capabilities['chrome']['userDataDir']}/config.json")
+        )
+        # ignore updates.
+        ot_application.config["alerts"]["ignored"] = ["appUpdateAvailable"]
+        ot_application.write_config()
+        robots_list = RobotsList(driver)
+        if not robots_list.is_robot_toggle_active(RobotsList.DEV):
+            robots_list.get_robot_toggle(RobotsList.DEV).click()
+        left_menu = LeftMenu(driver)
```
Comment (shlokamin) [inline app-testing/tests/protocol_upload_test.py:211]: is `LeftMenu` supposed to be the nav bar? if so can we name it `NavBar` so it matches our react component naming? doesnt matter a ton cuz we'll have to change this after 5.1 but consistency is nice

File: app-testing/tests/protocol_upload_test.py [hunk 14]
```diff
@@ -192,6 +188,97 @@ def test_moam_pur(
...
+        # Instantiate the page object for the RobotsList.
+        robots_list = RobotsList(driver)
+        # toggle the DEV robot
+        if not robots_list.is_robot_toggle_active(RobotsList.DEV):
+            robots_list.get_robot_toggle(RobotsList.DEV).click()
+        left_menu = LeftMenu(driver)
```
Comment (shlokamin) [inline app-testing/tests/protocol_upload_test.py]: how come this line is repeated? 

```
left_menu = LeftMenu(driver)
```

File: app-testing/tests/protocol_upload_test.py [hunk 15]
```diff
@@ -192,6 +188,97 @@ def test_moam_pur(
...
+        left_menu.click_protocol_upload_button()
+        protocol_file = ProtocolFile(driver)
+        logger.info(
+            f"uploading protocol: {test_protocols['protocoluploadjson'].resolve()}"
+        )
+        input = protocol_file.get_drag_json_protocol()
+        drag_and_drop_file(input, test_protocols["protocoluploadjson"])
```
Comment (shlokamin) [inline app-testing/tests/protocol_upload_test.py:223]: it looks like the protocol file for `protocoluploadjson` has no commands. don't we want to test LPC on a real protocol?

File: app-testing/tests/protocol_upload_test.py [hunk 16]
```diff
@@ -192,6 +188,97 @@ def test_moam_pur(
...
+            labware_position_check.get_deckmap_labware_check_complete().is_displayed()
+        )
+        assert labware_position_check.get_section_list_step0().is_displayed()
+        assert labware_position_check.get_section_list_step1().is_displayed()
+        assert labware_position_check.get_section_list_step2().is_displayed()
+        assert (
+            labware_position_check.get_close_and_apply_labware_offset_data_button().is_displayed()
+        )
+        labware_position_check.click_get_close_and_apply_labware_offset_data_button()
+        assert labware_position_check.get_labware_success_toast().is_displayed()
```
Comment (shlokamin) [inline app-testing/tests/protocol_upload_test.py:277]: it doesn't look like we ever validate the contents of the summary screen (i.e. what offsets we're about to apply. ex, if we jog twice to the left by 1mm on a piece of labware, we'd expect to see this offset show up in the summary screen. that's probably an important thing to test right?

Comment (shlokamin) [reply app-testing/tests/protocol_upload_test.py:277]: also, should we be testing that after we apply offsets that they show up in the labware setup step?

File: app-testing/src/pages/moam_pur.py (hunk 9 above)
Comment (reachojhaneha) [reply app-testing/src/pages/moam_pur.py]: this is done

File: app-testing/src/pages/module_setup.py (hunk 10 above)
Comment (reachojhaneha) [reply app-testing/src/pages/module_setup.py]: this is changed to toggle

File: app-testing/src/pages/module_setup.py (hunk 11 above)
Comment (reachojhaneha) [reply app-testing/src/pages/module_setup.py:89]: i have added the docstring

File: app-testing/files/protocol/python1/moam_pur.json (hunk 12 above)
Comment (reachojhaneha) [reply app-testing/files/protocol/python1/moam_pur.json:4]: moved the protocols to json folder

File: app-testing/tests/protocol_upload_test.py (hunk 13 above)
Comment (reachojhaneha) [reply app-testing/tests/protocol_upload_test.py:211]: a lot of this code will be rewritten with 5.1, as LeftMenu is being used by other tests out of PUR, i am not in favor of changing the LeftMenu class

File: app-testing/tests/protocol_upload_test.py (hunk 14 above)
Comment (reachojhaneha) [reply app-testing/tests/protocol_upload_test.py]: this is corrected

File: app-testing/tests/protocol_upload_test.py (hunk 15 above)
Comment (reachojhaneha) [reply app-testing/tests/protocol_upload_test.py:223]: this protocol has commands

File: app-testing/tests/protocol_upload_test.py (hunk 16 above)
Comment (reachojhaneha) [reply app-testing/tests/protocol_upload_test.py:277]: So, we decided to test the deckmap offset settings which inturn tests the summary screen. the asserts have been added to the tests

**Review by shlokamin — APPROVED** [review]
Approved pending changing DOM ids to be unique and validating labware offsets for ALL labware (not just the first one)

**Review by y3rsh — APPROVED** [review]
Approved with the same caveat as @shlokamin

**Comment by reachojhaneha** [comment]
Now, the test checks all the labware offsets in the PR

**Pull request merged**